ninja-core/src/main/java/ninja/ReverseRouter.java

```diff
@@ -0,0 +1,421 @@
+         * parameters it's recommended to use the <code>path</code> method since
+         * it will validate the parameter exists, while this method would fallback
+         * to adding the value as a query string parameter if the path parameter
+         * did not exist.
+         * @param parameterMap The list of pairs of name and values.
+         * @return A reference to this builder
+         * @see #path(java.lang.String, java.lang.Object)
+         * @see #query(java.lang.String, java.lang.Object) 
+         */
+        public Builder rawParams(Object... parameterMap) {
+            return setParams(true, parameterMap);
+        }
+        
+        private Builder setParams(boolean raw, Object... parameterMap) {
+            if (parameterMap != null) {
+                if (parameterMap.length % 2 != 0) {
+                    throw new IllegalArgumentException("Always provide key (as String) value (as Object) pairs in parameterMap. That means providing e.g. 2, 4, 6... objects.");
+                }
+
+                for (int i = 0; i < parameterMap.length; i += 2) {
+                    setParam((String)parameterMap[i], parameterMap[i+1], raw);
+                }
+            }
+            return this;
+        }
+        
+        /**
+         * Add a parameter as a path replacement if it exists or will fallback
+         * to adding it a a query string value.  This method will URL encode the
+         * values when building the final result.
+         * @param name The parameter name
+         * @param value The parameter value
+         * @return A reference to this builder
+         * @see #path(java.lang.String, java.lang.Object)
+         * @see #query(java.lang.String, java.lang.Object) 
+         */
+        public Builder param(String name, Object value) {
+            return setParam(name, value, false);
+        }
+        
+        /**
+         * Identical to <code>param</code> except this method will NOT url
+         * encode the values when building the final result.
+         * @param name The parameter name
+         * @param value The parameter value
+         * @return A reference to this builder
+         * @see #path(java.lang.String, java.lang.Object)
+         * @see #query(java.lang.String, java.lang.Object)
+         */
+        public Builder rawParam(String name, Object value) {
+            return setParam(name, value, true);
+        }
+        
+        private Builder setParam(String name, Object value, boolean raw) {
+            Objects.requireNonNull(name, "name required");
+            
+            if (this.route.getParameters().containsKey(name)) {
+                return setPath(name, value, raw);
+            } else {
+                return setQuery(name, value, raw);
+            }
+        }
+        
+        private String safeValue(Object value, boolean raw) {
+            String s = (value == null ? null : value.toString());
```

Comment:
The query parameter value can be null -- where its just in the query string but doesn't have the "=value" part on it.